src/ProcessManager.php

```diff
@@ -263,176 +200,98 @@ public function shutdown($graceful = true)
     {
-        $this->appenv = $appenv;
-    }
+        if (array_key_exists($eventId, $this->onceListeners)) {
+            foreach ($this->onceListeners[$eventId] as $listener) {
+                $listener();
+            }
 
-    /**
-     * @return ?string
-     */
-    public function getAppEnv()
-    {
-        return $this->appenv;
-    }
+            $this->onceListeners[$eventId] = [];
+        }
 
-    /**
-     * @return boolean
-     */
-    public function isLogging()
-    {
-        return $this->logging;
+        if (array_key_exists($eventId, $this->listeners)) {
+            foreach ($this->listeners[$eventId] as $listener) {
+                $listener();
+            }
+        }
     }
 
     /**
-     * @param boolean $logging
+     * To be called after all workers have been terminated and the event loop is no longer in use.
      */
-    public function setLogging($logging)
+    protected function quit()
     {
-        $this->logging = $logging;
-    }
+        $this->output->writeln('Stopping the process manager.');
 
-    /**
-     * @return string
-     */
-    public function getStaticDirectory()
-    {
-        return $this->staticDirectory;
-    }
+        // this method is also called during startup when something crashed, so
+        // make sure we don't operate on nulls.
+        if ($this->controller) {
+            @$this->controller->close();
+        }
+        if ($this->web) {
+            @$this->web->close();
+        }
 
-    /**
-     * @param string $staticDirectory
-     */
-    public function setStaticDirectory($staticDirectory)
-    {
-        $this->staticDirectory = $staticDirectory;
-    }
+        if ($this->loop) {
+            $this->loop->tick();
+            $this->loop->stop();
+        }
 
-    public function setPIDFile($pidfile)
-    {
-        $this->pidfile = $pidfile;
-    }
-    /**
-     * @return boolean
-     */
-    public function isDebug()
-    {
-        return $this->debug;
+        if (file_exists($this->config->getPIDFile())) {
+            unlink($this->config->getPIDFile());
+        }
+        exit;
     }
 
     /**
-     * @param boolean $debug
+     * Manage the hot code reload timer as per the debug setting. Should be called during a reload.
      */
-    public function setDebug($debug)
+    protected function updateCodeReloadTimer()
     {
-        $this->debug = $debug;
-    }
+        // Update the state of the debug hot code reload timer
+        if ($this->config->isDebug() && !$this->debugTimer) {
+            if ($this->output->isVeryVerbose()) {
+                $this->output->writeln('Debug mode is active: hot code reloading is now enabled.');
+            }
 
-    /**
-     * @return int
-     */
-    public function getReloadTimeout()
-    {
-        return $this->reloadTimeout;
-    }
+            $this->debugTimer = $this->loop->addPeriodicTimer(0.5, function () {
+                $this->checkChangedFiles();
```

Comment:
I think we can change this whole part with the same mechanisms as suggested in https://github.com/php-pm/php-pm/issues/376 - wdyt? It's a separate topic though.

Reply:
Sure. But in another PR :^)